lib/Frontend/FrontendInputsAndOutputs.cpp

```diff
@@ -238,3 +256,90 @@ void FrontendInputsAndOutputs::addPrimaryInputFile(StringRef file,
                                                    llvm::MemoryBuffer *buffer) {
   addInput(InputFile(file, true, buffer));
 }
+
+// Outputs
+
+unsigned FrontendInputsAndOutputs::countOfInputsProducingMainOutputs() const {
+  return isSingleThreadedWMO()
+             ? 1
+             : hasPrimaryInputs() ? primaryInputCount() : inputCount();
+}
+
+const InputFile &FrontendInputsAndOutputs::firstInputProducingOutput() const {
+  return isSingleThreadedWMO()
+             ? firstInput()
+             : hasPrimaryInputs() ? firstPrimaryInput() : firstInput();
+}
+
+const InputFile &FrontendInputsAndOutputs::lastInputProducingOutput() const {
+  return isSingleThreadedWMO()
+             ? firstInput()
+             : hasPrimaryInputs() ? lastPrimaryInput() : lastInput();
+}
+
+void FrontendInputsAndOutputs::forEachInputProducingAMainOutputFile(
+    llvm::function_ref<void(const InputFile &)> fn) const {
+  isSingleThreadedWMO()
+      ? fn(firstInput())
+      : hasPrimaryInputs() ? forEachPrimaryInput(fn) : forEachInput(fn);
+}
+
+void FrontendInputsAndOutputs::setMainOutputs(
+    ArrayRef<std::string> outputFiles) {
+  assert(countOfInputsProducingMainOutputs() == outputFiles.size());
+  if (hasPrimaryInputs()) {
+    unsigned i = 0;
+    for (auto index : indices(AllInputs)) {
+      InputFile &f = AllInputs[index];
+      if (f.isPrimary())
+        f.setOutputFilename(outputFiles[i]);
```

Comment:
You never increment `i` here.

Reply:
Thought I had typed the "++"! But fixed, now, tx.